floodviz/utils.py

```diff
@@ -0,0 +1,90 @@
+            # should run once. Used to map column headers to line split indices.
+            if not found_headers:
+                raw_headers = line.split('\t')
+                for idx, key in enumerate(raw_headers):
+                    if header_values.get(key) is not None:
+                        keep_headers[key] = idx
+                found_headers = True
+                continue
+            else:
+                if line.startswith('USGS'):
+                    line = line.split('\t')
+                    data_point = {}
+                    for key, value in keep_headers.items():
+                        data_point[key] = line[value]
+                    all_data.append(data_point)
+        content = all_data
+    return content
+
+""" 
+    Takes a web service endpoint, a dictionary of key/value pairs for the query parameters
+    and returns the formatted url. 
+
+    ARGS: 
+        endpoint - String specifiying the web service url
+        params - dictionary holding url parameters
+            
+            possible keys:
+                'sites' (mandatory) - List of string site_ids
+                'site_query' (mandatory) - service specific string for url construction that targets the sites.
+                'format' (mandatory) - service specific string for format of the data returned from service
+                'extra' - service specific string prepended to the format if needed.
+                'start_date_query' - service specific string for url construction that targets the start date of data
+                'start_date' (madatory with start_date_query) - String specifying the start date for data. (YYYY-MM-DD)
+                'end_date_query' - service specific string for url construction that targets the end date of data
+                'end_date' (madatory with end_date_query) - String specifying the end date for data. (YYYY-MM-DD)
+                'site_status' - service specific string for url construction. Cannot be used with 'agency' key
+                'agency' - service specific string for url construction. Cannot be used with 'site_status' key
+
+    RETURNS:
+        url - url to request for data 
+"""
+
+
+def url_construct(endpoint, params):
+    # Construct request URL
+    sites = ','.join(params['sites'])
+    url = endpoint 
+    if params.get('extra'):
+        url += params['extra'] + params['format']
+    else:
+        url += params['format']
+    if params.get('site_query'):
+        url += params['site_query'] + sites
+    if params.get('start_date_query'):
+        url += params['start_date_query'] + params['start_date']
+    if params.get('end_date_query'):
+        url += params['end_date_query'] + params['end_date']
+    # one or the other. Not both. 'site_status' and 'agency' keys should not be in the same params dict.
+    if params.get('site_status'):
+        url += params['site_status']
+    elif params.get('agency'):
+        url += params['agency']
+    return url
```

Comment:
I don't think this function is likely to make anything easier. There's very few common components between the URLs for all our different requests, so I think it'll be easier to handle URL construction on a case-by-case basis. Constructing the URLs this way would make more sense if we were making a bunch of calls with very similar parameters.